### ISSUE TITLE
Glass? .. Can anyone explain how to do it?`

### ISSUE THREAD

**Issue by SparkieShock** [KhronosGroup/glTF-Blender-Exporter#319]
Been trying this all night with a variety of options but none has yet produced as single transparent surface. Can one not do glass with GLB yet?

**Comment by donmccurdy** [comment]
I'd refer to the docs and images here: https://github.com/KhronosGroup/glTF-Blender-Exporter/blob/master/docs/user.md#pbr-materials. The `Alpha` socket accepts an `Image Texture` node (and nothing else, I think). The `BaseColorFactor` alpha channel can be used to manually adjust alpha. If this isn't working, could you share a screenshot of your material configuration?

glTF does not currently support physically-based glass materials, just alpha blending.

**Comment by SparkieShock** [comment]
Well .. I appreciate the answers .. sadly none of this worked. I'm just gonna assume for the meantime that glass or transparent materials for oculus glb models isn't possible. Look forward to when it is.

**Comment by emackey** [comment]
@SparkieShock Take a look at the [Alpha Blend Mode Test here](https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/AlphaBlendModeTest).  If your Oculus Glb viewer can't render it as described in the model's readme, then that's a bug that should be filed on the viewer software.  (Also if there's any way for you to post a screenshot, I'd be curious to see).

**Comment by ankervold** [comment]
@SparkieShock  I'm a beginner and had the same issue with as you with trying to find info on creating glass or transparent material but found this thread helpful http://www.html5gamedevs.com/topic/38227-gltf-dont-apply-alphatransparency-solved/. 
Here's one of my setups that work.
![screenshot 2](https://user-images.githubusercontent.com/42947756/46828036-2d978e00-cd68-11e8-8e27-0e184bd178ce.png)
 I think this is what @donmccurdy was referring to in the first reply.
The BaseColorFactor has R B G A factors, set RBG to 1.0 but set the A between 0 and 1, I don't think it matters what as long as it less than 1. Then your Alpha slider will allow you to adjust the transparency, 0 completely see through 1 opaque. Keep AlphaCutoff 0.5. Metallic factor and Roughness factor set to 0.

**Comment by SparkieShock** [comment]
@ankervold tried your setup .. didn't work but then it might be due to my not using the same image texture as what you have used. What did you use?
Otherwise my glass tube is still just a big ol' block of white.

**Comment by ankervold** [comment]
@SparkieShock My image texture is just a simple color palette. I don't think it should matter though I used a wood texture and still got it to be transparent. Check to make sure you have all the inputs right, I know I changed the BaseColor A (alpha) channel rather than the BaseColorFactor A (alpha) channel when first trying it out. I was trying out MetallicFactor and RoughnessFactor and they don't have to be 0 so it's really just changing the BaseColorFactor A channel to less than 1 and using the Alpha slider (below the EmissiveFactor) to control how transparent you want it.

**Comment by SparkieShock** [comment]
@ankervold I tried using a pure white texture .. I tried using a blank (alpha) texture .. the only difference is that the white texture means the glass tube comes out white .. and with the alpha it comes out as black .. both are with reflections .. but I've tried changing the alpha input value from 0,1 all the way to 0.8 and there's no difference. Upon importing to oculus it just won't go transparent.

**Comment by SparkieShock** [comment]
Okay so I got the model to sort of work. In the windows 3D model viewer the glass looks pretty good .. sadly in oculus it kinda looks like there's VHS static moving around the glass tube .. so it kinda worked .. but doesn't really look like glass in oculus.

Here is a screen shot of the Shaders
![capture 2](https://user-images.githubusercontent.com/30974358/46860930-49318180-cdc6-11e8-9ace-ee67285aaf1c.JPG)

I used a texture which was two layers .. transparent background with a white layer on top but reduced to 20% opacity. And then using the flow as above.

And the result:
![tardis with glass](https://user-images.githubusercontent.com/30974358/46860997-69f9d700-cdc6-11e8-9c88-2fd8c7fd72ff.JPG)

Looks okay-ish in VR .. but not like it does here.

**Comment by ankervold** [comment]
So from looking at this I think there's two way to make transparent materials. The first is how you and I think the Polly example linked above do it and that is to have a multilayered texture with some parts being transparent. Then hooking the image texture in the BaseColor and Alpha nodes you get the texture outputting a similar transparency you made for it beforehand. 

The second way to do it is have the image texture only connect to the BaseColor then change the BaseColorFactor A channel to less than 1.0. Then you can use the Alpha slider further down to change the transparency. The downside to this is it will change the transparency of the entire texture so if you have parts you want more opaque you probably have to use the first method. I've been using the second one but my scenes have been low detail single color materials so I haven't needed to layer a texture like that.

The VHS noise I don't know how to get rid of unless you made a specific texture that acted like glass, I haven't tried anything like that. Its easy enough to fake a transparent material to stand in for glass but I have no idea how to make something that could actually pass for glass.

In the shader screenshot you link I don't think you need an RGB node attached to Alphamode. Also the BaseColorFactor is still 1.0 but with your method it still works.

Also I'm using a Lenovo Explorer so if it's an oculus specific issue I can't really help.

**Comment by SparkieShock** [comment]
Yeah the RGB thing hooked to alpha actually did nothing .. but after rendering about 27 versions of this model THIS was the first setup that actually worked .. kinda .. sorta!

I'm afraid I don't really understand what you wrote about the first method .. all that really went over my head. Not sure why you'd need a multilayered texture, or even how you'd do that .. 2 image texture nodes? .. but only one can be hooked up at one time so how would you do two?

The sad thing is that in the model looks great in the windows 3D model viewer .. so clearly this idea of mine worked .. BUT! .. for it's intended destination (Oculus and VR in general) it looks very hap hazzard and not at all like glass. Of course this is a limit of the technology within Oculus .. and it's all still very much beta software over there so perhaps this stuff will improve. Perhaps I'll send my model and concerns to them so they can advise or at least study the problem.

It's just a bloody shame that you can't link a 'transparent' shader to an output and be done ... all this dicking around with alpha layers .. metallic nodes and stuff is headache inducing.

**Comment by ankervold** [comment]
Yea sorry it's more complicated then I let on lol. https://www.khronos.org/blog/art-pipeline-for-gltf Is where I learned about combining textures. You would have to use an image editing software like GIMP or photoshop to combine the textures then import them into blender.

**Comment by SparkieShock** [comment]
Ah right .. well thats fine .. I mean I had to use Afinity in order to produce a layering of white and transparent background .. so this is actually the workflow I used anyways.

I think it's safe to say that at present 'glass' material just isn't possible for none official oculus objects. Clearly they have some proprietary software to manage their shaders .. cuz it sure as hell isn't working under glb. There's no good reason my model should look find under a model viewer and not work properly in oculus. (Maybe there's like some kind of multiple layer calculation going on that's causing that interference pattern .. I noticed that my graphics card is working over time to render the glass tube ... turn it off and all of a sudden my GTX1080ti is hardly doing any work at all).

Still .. thanks for everyone's help. Hopefully after a bit more time for development there will be much easier glass shader possiblities for VR objects.
